crates/sui-core/src/transaction_input_loader.rs

```diff
@@ -0,0 +1,349 @@
+                        Ok::<HashMap<ObjectID, SequenceNumber>, SuiError>(
+                            shared_lock_store
+                                .get_shared_locks(tx_digest)?
+                                .into_iter()
+                                .collect(),
+                        )
+                    })?;
+                    // If we can't find the locked version, it means
+                    // 1. either we have a bug that skips shared object version assignment
+                    // 2. or we have some DB corruption
+                    let version = shared_locks.get(id).unwrap_or_else(|| {
+                        panic!(
+                            "Shared object locks should have been set. tx_digest: {:?}, obj id: {:?}",
+                            tx_digest, id
+                        )
+                    });
+                    object_keys.push(ObjectKey(*id, *version));
+                    fetch_indices.push(i);
+                }
+            }
+        }
+
+        let objects = self.store.multi_get_object_by_key(&object_keys)?;
+
+        for (object, input, key, index) in izip!(
+            objects.into_iter(),
+            input_object_kinds,
+            object_keys.into_iter(),
+            fetch_indices.into_iter()
+        ) {
+            results[index] = Some(match (object, input) {
+                (Some(obj), input_object_kind) => ObjectReadResult {
+                    input_object_kind: *input_object_kind,
+                    object: obj.into(),
+                },
+                (None, InputObjectKind::SharedMoveObject { id, .. }) => {
+                    // Check if the object was deleted by a concurrently certified tx
+                    let version = key.1;
+                    if let Some(dependency) = self.store.get_deleted_shared_object_previous_tx_digest(id, &version, epoch_id)? {
+                        ObjectReadResult {
+                            input_object_kind: *input,
+                            object: ObjectReadResultKind::DeletedSharedObject(version, dependency),
+                        }
+                    } else {
+                        panic!("All dependencies of tx {:?} should have been executed now, but Shared Object id: {}, version: {} is absent in epoch {}", tx_digest, *id, version, epoch_id);
+                    }
+                },
+                _ => panic!("All dependencies of tx {:?} should have been executed now, but obj {:?} is absent", tx_digest, key),
+            });
+        }
+
+        Ok(results
+            .into_iter()
+            .map(Option::unwrap)
+            .collect::<Vec<_>>()
+            .into())
+    }
+}
+
+// private methods
+impl TransactionInputLoader {
+    async fn read_objects_for_synchronous_execution_impl(
+        &self,
+        _tx_digest: Option<&TransactionDigest>,
+        input_object_kinds: &[InputObjectKind],
+        receiving_objects: &[ObjectRef],
```

Comment:
we check the length to enforce max objects limits. Also, as I was discussing with @tzakian, we may want to start checking receiving objects during dry-run exec, in which case this will most likely be used

Reply:
That's not a good argument to leave this not processed logically. We could move the fp_ensure check to the two callers for now until we decide to implement it

Reply:
as i said in the other comment, it turns out this can (and should) already be used for dry run exec